hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,260 @@
 
     def set_exception(self, exception: BaseException):
-        self._sync_updates()
-        if self._state in self.TERMINAL_STATES:
-            raise FutureStateError(f"Can't set_exception to a future that is {self._state} ({self})")
-        self._state, self._exception = base.FINISHED, exception
-        self._send_updates()
+        if os.getpid() == self._origin_pid:
+            self._active_futures.pop(self._uid, None)
+            super().set_exception(exception)
+        elif self._state in TERMINAL_STATES:
+            raise InvalidStateError(f"Can't set_exception to a future that is {self._state} ({self._uid})")
+        else:
+            self._state_cache[self._state], self._exception = base.FINISHED, exception
+            self._send_update(UpdateType.EXCEPTION, exception)
+
+    def cancel(self) -> bool:
+        if os.getpid() == self._origin_pid:
+            self._active_futures.pop(self._uid, None)
+            return super().cancel()
+        elif self._state in [base.RUNNING, base.FINISHED]:
+            return False
+        else:
+            self._state_cache[self._state] = base.CANCELLED
+            self._send_update(UpdateType.CANCEL)
+            return True
 
     def set_running_or_notify_cancel(self):
-        self._sync_updates()
         if self._state == base.PENDING:
             self._state = base.RUNNING
-            return self._send_updates()
+            return True
         elif self._state == base.CANCELLED:
             return False
         else:
-            raise FutureStateError(f"Can't set_running_or_notify_cancel to a future that is in {self._state} ({self})")
-
-    def cancel(self):
-        self._sync_updates()
-        if self._state in self.TERMINAL_STATES:
-            return False
-        self._state, self._exception = base.CANCELLED, base.CancelledError()
-        return self._send_updates()
+            raise InvalidStateError(f"Can't set_running_or_notify_cancel when future is in {self._state} ({self._uid})")
 
     def result(self, timeout: Optional[float] = None) -> ResultType:
-        self._await_terminal_state(timeout)
-        if self._exception is not None:
+        if self._state not in TERMINAL_STATES:
+            assert os.getpid() == self._origin_pid, "only the process that created MPFuture can await result."
+            return super().result(timeout)
+        elif self._state == base.CANCELLED:
+            raise base.CancelledError()
+        elif self._exception:
             raise self._exception
-        return self._result
+        else:
+            return self._result
 
-    def exception(self, timeout=None) -> BaseException:
-        self._await_terminal_state(timeout)
-        if self._state == base.CANCELLED:
+    def exception(self, timeout: Optional[float] = None) -> BaseException:
+        if self._state not in TERMINAL_STATES:
+            assert os.getpid() == self._origin_pid, "only the process that created MPFuture can await exception."
```

Comment:
IMHO, in such cases raising a specific error is [better](https://stefan.sofa-rockers.org/2018/04/16/assertions-and-exceptions/) than just an assertion. The motivation is that this is an external method and errors can happen [on the user side](https://stackoverflow.com/a/54252850) (basically we are trying to validate the environment), second reason is that it will be more informative and specific if one decides to handle this error.

Reply:
thx, fixed